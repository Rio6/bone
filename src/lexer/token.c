#include "token.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char * const TokenType_names[] = {
   "T_ERROR",
   "T_IDENT",
   "T_KEYWORD",
   "T_INT",
   "T_FLOAT",
   "T_CHAR",
   "T_STRING",
   "T_COMMENT",
   "T_BRACE_L",
   "T_BRACE_R",
   "T_PAREN_L",
   "T_PAREN_R",
   "T_BRACK_L",
   "T_BRACK_R",
   "T_DOT",
   "T_COMMA",
   "T_SEMICOLON",
   "T_COLON",
   "T_REF",
   "T_DEREF",
   "T_PLUS",
   "T_MINUS",
   "T_MULT",
   "T_DIV",
   "T_MOD",
   "T_LT",
   "T_GT",
   "T_AND",
   "T_OR",
   "T_XOR",
   "T_NOT",
   "T_ASSIGN",
   "T_LSHIFT",
   "T_RSHIFT",
   "T_EQ",
   "T_NE",
   "T_LE",
   "T_GE",
   "T_ASPLUS",
   "T_ASMINUS",
   "T_ASMULT",
   "T_ASDIV",
   "T_ASMOD",
   "T_ASAND",
   "T_ASOR",
   "T_ASXOR",
   "T_ASLSHIFT",
   "T_ASRSHIFT",
   "T_EOT",
};

Token *token_create(TokenType type, const char *value) {
   char *copied_value = NULL;
   if(value) {
      size_t len = strlen(value) + 1;
      copied_value = malloc(len);
      strcpy(copied_value, value);
   }

   return token_create_raw(type, copied_value);
}

Token *token_create_raw(TokenType type, char *value) {
   Token *token = calloc(1, sizeof(Token));
   token->type = type;
   token->value = value;
   return token;
}

void token_delete(Token *token) {
   if(!token) return;
   token_delete(token->next);
   free(token->value);
   free(token);
}

void token_dump(Token *token) {
   if(!token) return;
   token_print(token);
   token_dump(token->next);
}

void token_print(Token *token) {
   printf("type: %s, value: %s\n", TokenType_names[token->type], token->value);
}
