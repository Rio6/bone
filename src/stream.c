#include "stream.h"

#include <string.h>

char stream_getc(Stream *s) {
   if(s->ahead > 0) {
      char c = s->buff[s->pos++];
      s->pos %= sizeof(s->buff);
      s->ahead--;
      return c;
   } else {
      char c = fgetc(s->stream);
      s->buff[s->pos++] = c;
      s->pos %= sizeof(s->buff);
      return c;
   }
}

char stream_peak(Stream *s) {
   if(s->ahead > 0) {
      return s->buff[s->pos];
   } else {
      char c = fgetc(s->stream);
      ungetc(c, s->stream);
      return c;
   }
}

char stream_ungetc(Stream *s, size_t count) {
   s->pos = (s->pos - count + sizeof(s->buff)) % sizeof(s->buff); // s->pos - count, but with modular math
   s->ahead += count;
   return s->buff[s->pos];
}

void stream_begins(Stream *s) {
   s->begin = s->pos;
}

char *stream_ends(Stream *s, char *buff, size_t size) {
   size_t count = 0;

   for(size_t i = s->begin; i != s->pos && count < size; i = (i+1) % sizeof(s->buff), count++) {
      buff[count] = s->buff[i];
   }

   if(count >= size) {
      buff[size-1] = '\0';
   } else {
      buff[count] = '\0';
   }

   return buff;
}

void stream_rollback(Stream *s) {
   ssize_t rollback_size = s->pos - s->begin;

   if(rollback_size > 0) {
      s->ahead += rollback_size;
   } else {
      s->ahead += sizeof(s->buff) + rollback_size;
   }

   s->pos = s->begin;
}
