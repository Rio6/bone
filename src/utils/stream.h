#pragma once

#include <stdio.h>

#define STREAM_BUFF 1024

typedef struct {
   FILE *stream;
   char buff[STREAM_BUFF];
   size_t pos, ahead, begin;
} Stream;

char stream_getc(Stream*);
char stream_peek(Stream*);
char stream_ungetc(Stream*, size_t);
void stream_begins(Stream*);
char *stream_ends(Stream*, char*, size_t);
void stream_rollback(Stream*);
