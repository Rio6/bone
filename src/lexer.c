#include "lexer.h"
#include "stream.h"

#include <stdlib.h>
#include <string.h>
#include <ctype.h>

static Lexer * const lexers[] = {
   comment_lexer,
   number_lexer,
   string_lexer,
   ident_lexer,
   sep_lexer,
   assign_lexer,
   operator_lexer,
};

Token *lex(Stream *stream) {
   Token head_token = {0};
   Token *last_token = &head_token;

   while(1) {
      int has_token = 0;
      char last_c = '\0';

      for(size_t i = 0; i < sizeof(lexers) / sizeof(lexers[0]); i++) {
         // advance until not a white space
         while(isspace(stream_getc(stream)));
         last_c = stream_ungetc(stream, 1);

         Token *token = lexers[i](stream);
         if(token) {
            last_token->next = token;
            last_token = token;
            has_token = 1;
            break;
         }
      }

      if(!has_token) {
         if(last_c != EOF) {
            fprintf(stderr, "Unexpected character: %1$c (%1$X)\n", last_c);
         }
         break;
      }
   }

   return head_token.next;
}

Token *token_create(TokenType type, const char *value) {
   Token *token = calloc(1, sizeof(Token));

   token->type = type;

   if(value) {
      size_t len = strlen(value) + 1;
      token->value = malloc(len);
      memcpy(token->value, value, len);
   }

   return token;
}

void token_delete(Token *token) {
   if(!token) return;
   token_delete(token->next);
   free(token->value);
   free(token);
}

void token_dump(Token *token) {
   if(!token) return;
   printf("type: %s, value: %s\n", token_type_name(token->type), token->value);
   token_dump(token->next);
}

char *token_type_name(TokenType type) {
   switch(type) {
      case IDENT:   return "IDENT";
      case KEYWORD: return "KEYWORD";
      case INT:     return "INT";
      case FLOAT:   return "FLOAT";
      case CHAR:    return "CHAR";
      case STRING:  return "STRING";
      case COMMENT: return "COMMENT";
      case BLOCK_L: return "BLOCK_L";
      case BLOCK_R: return "BLOCK_R";
      case PAREN_L: return "PAREN_L";
      case PAREN_R: return "PAREN_R";
      case LIST_L:  return "LIST_L";
      case LIST_R:  return "LIST_R";
      case DOT:     return "DOT";
      case COMMA:   return "COMMA";
      case SEMICOLON: return "SEMICOLON";
      case ASSIGN:  return "ASSIGN";
      case OP:      return "OP";
      case REF:     return "REF";
      case DEREF:   return "DEREF";
      default:      return "UNKNOWN";
   }
}
