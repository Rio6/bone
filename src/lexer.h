#pragma once

#include "stream.h"

typedef enum {
   ERROR,
   IDENT,
   KEYWORD,
   INT,
   FLOAT,
   CHAR,
   STRING,
   COMMENT,
   BLOCK_L,
   BLOCK_R,
   PAREN_L,
   PAREN_R,
   LIST_L,
   LIST_R,
   DOT,
   COMMA,
   SEMICOLON,
   ASSIGN,
   OP,
   REF,
   DEREF,
} TokenType;

typedef struct Token {
   TokenType type;
   char *value; // owner
   struct Token *next; // owner
} Token;

typedef Token *(Lexer)(Stream *stream);

extern Lexer *lexers[];
extern const char *TokenType_names[];

Token *lex(Stream *stream);

Token *token_create(TokenType type, const char *value);
Token *token_create_raw(TokenType type, char *value);
void token_delete(Token*);
void token_dump(Token*);

char *token_type_name(TokenType);

Token *ident_lexer(Stream*);
Token *sep_lexer(Stream*);
Token *number_lexer(Stream*);
Token *char_lexer(Stream*);
Token *string_lexer(Stream*);
Token *comment_lexer(Stream*);
Token *assign_lexer(Stream*);
Token *operator_lexer(Stream*);
