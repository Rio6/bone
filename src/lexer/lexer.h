#pragma once

#include "token.h"
#include "utils/stream.h"

typedef Token *(Lexer)(Stream *stream);

extern Lexer *lexers[];

Token *lexer_run(Stream *stream);
