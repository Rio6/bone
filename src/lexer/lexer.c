#include "lexer.h"
#include "token.h"

#include "utils/stream.h"

#include <string.h>
#include <ctype.h>

Token *lexer_run(Stream *stream) {
   Token head_token = {0};
   Token *last_token = &head_token;

   while(1) {
      int has_token = 0;
      char last_c = '\0';

      for(Lexer **lexer = lexers; *lexer != NULL; lexer++) {

         // advance until not a white space
         while(isspace(stream_getc(stream)));
         last_c = stream_ungetc(stream, 1);

         Token *token = (*lexer)(stream);

         if(token) {
            last_token->next = token;
            last_token = token;
            has_token = 1;
            break;
         }
      }

      if(!has_token) {
         char buff[128];
         snprintf(buff, sizeof(buff), "Unexpected character: %1$c (%1$X)\n", last_c);
         last_token->next = token_create(T_ERROR, buff);
         last_token = last_token->next;
      }

      if(last_token->type == T_ERROR || last_token->type == T_EOT) {
         break;
      }
   }

   return head_token.next;
}
