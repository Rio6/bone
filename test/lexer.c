#include "lexer/lexer.h"
#include "utils/stream.h"

#include <stdio.h>

#define TEST_FILE "../idea.bone"

int main(void) {
   FILE *file = fopen(TEST_FILE, "r");

   if(!file) {
      perror("error open file");
      return 1;
   }

   Stream stream = {
      .stream = file
   };

   Token *tokens = lexer_run(&stream);
   fclose(file);
   token_dump(tokens);
   token_delete(tokens);

   return 0;
}
