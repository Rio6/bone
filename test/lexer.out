type: T_KEYWORD, value: import
type: T_IDENT, value: std
type: T_DOT, value: (null)
type: T_IDENT, value: math
type: T_DOT, value: (null)
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: import
type: T_IDENT, value: std
type: T_DOT, value: (null)
type: T_IDENT, value: io
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: import
type: T_DOT, value: (null)
type: T_IDENT, value: directory
type: T_DOT, value: (null)
type: T_IDENT, value: file
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: import
type: T_DOT, value: (null)
type: T_IDENT, value: directory
type: T_DOT, value: (null)
type: T_STRING, value: fancy-name
type: T_DOT, value: (null)
type: T_IDENT, value: name
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: export
type: T_KEYWORD, value: struct
type: T_IDENT, value: A
type: T_BRACE_L, value: (null)
type: T_IDENT, value: i32
type: T_IDENT, value: a
type: T_SEMICOLON, value: (null)
type: T_COMMENT, value: (null)
type: T_IDENT, value: i32
type: T_IDENT, value: b
type: T_ASSIGN, value: (null)
type: T_INT, value: 10
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: f64
type: T_IDENT, value: c
type: T_ASSIGN, value: (null)
type: T_FLOAT, value: 10.23
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: f32
type: T_IDENT, value: d
type: T_ASSIGN, value: (null)
type: T_FLOAT, value: .65
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: f32
type: T_IDENT, value: e
type: T_ASSIGN, value: (null)
type: T_FLOAT, value: .4e10
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: f32
type: T_IDENT, value: f
type: T_ASSIGN, value: (null)
type: T_FLOAT, value: 4e4
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: char
type: T_PAREN_L, value: (null)
type: T_IDENT, value: A
type: T_IDENT, value: self
type: T_COMMA, value: (null)
type: T_IDENT, value: bool
type: T_IDENT, value: stuff
type: T_PAREN_R, value: (null)
type: T_REF, value: (null)
type: T_IDENT, value: method
type: T_ASSIGN, value: (null)
type: T_BRACE_L, value: (null)
type: T_KEYWORD, value: return
type: T_CHAR, value: x
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: char
type: T_PAREN_L, value: (null)
type: T_IDENT, value: A
type: T_COMMA, value: (null)
type: T_IDENT, value: bool
type: T_PAREN_R, value: (null)
type: T_REF, value: (null)
type: T_IDENT, value: sameThing
type: T_ASSIGN, value: (null)
type: T_IDENT, value: char
type: T_PAREN_L, value: (null)
type: T_IDENT, value: A
type: T_IDENT, value: self
type: T_COMMA, value: (null)
type: T_IDENT, value: bool
type: T_IDENT, value: stuff
type: T_PAREN_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_BRACE_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_IDENT, value: i32
type: T_REF, value: (null)
type: T_IDENT, value: aPtr
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: i32
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_IDENT, value: array
type: T_ASSIGN, value: (null)
type: T_BRACE_L, value: (null)
type: T_INT, value: 1
type: T_COMMA, value: (null)
type: T_INT, value: 2
type: T_COMMA, value: (null)
type: T_INT, value: 3
type: T_COMMA, value: (null)
type: T_INT, value: 4
type: T_BRACE_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: i32
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_IDENT, value: arrayAgain
type: T_ASSIGN, value: (null)
type: T_IDENT, value: i32
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_INT, value: 1
type: T_COMMA, value: (null)
type: T_INT, value: 2
type: T_COMMA, value: (null)
type: T_INT, value: 3
type: T_COMMA, value: (null)
type: T_INT, value: 4
type: T_BRACE_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: u8
type: T_BRACK_L, value: (null)
type: T_INT, value: 128
type: T_BRACK_R, value: (null)
type: T_IDENT, value: preallocArray
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: i32
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_REF, value: (null)
type: T_IDENT, value: ptrToArray
type: T_ASSIGN, value: (null)
type: T_REF, value: (null)
type: T_IDENT, value: array
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: i32
type: T_REF, value: (null)
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_IDENT, value: arrayOfPtr
type: T_ASSIGN, value: (null)
type: T_IDENT, value: null
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: i32
type: T_PAREN_L, value: (null)
type: T_IDENT, value: char
type: T_PAREN_R, value: (null)
type: T_REF, value: (null)
type: T_IDENT, value: ptrToFuncReturnsI32
type: T_SEMICOLON, value: (null)
type: T_COMMENT, value: (null)
type: T_IDENT, value: char
type: T_PAREN_L, value: (null)
type: T_IDENT, value: A
type: T_IDENT, value: byValue
type: T_COMMA, value: (null)
type: T_IDENT, value: A
type: T_REF, value: (null)
type: T_IDENT, value: byPointer
type: T_COMMA, value: (null)
type: T_IDENT, value: A
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_IDENT, value: byValue
type: T_COMMA, value: (null)
type: T_IDENT, value: A
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_REF, value: (null)
type: T_IDENT, value: byReferenceOfArray
type: T_PAREN_R, value: (null)
type: T_IDENT, value: aFunction
type: T_ASSIGN, value: (null)
type: T_BRACE_L, value: (null)
type: T_KEYWORD, value: if
type: T_PAREN_L, value: (null)
type: T_IDENT, value: byValue
type: T_DOT, value: (null)
type: T_IDENT, value: a
type: T_EQ, value: (null)
type: T_IDENT, value: byPointer
type: T_DOT, value: (null)
type: T_IDENT, value: b
type: T_PAREN_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_COMMENT, value: (null)
type: T_KEYWORD, value: return
type: T_IDENT, value: byPointer
type: T_DOT, value: (null)
type: T_IDENT, value: c
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_IDENT, value: A
type: T_IDENT, value: copied
type: T_ASSIGN, value: (null)
type: T_DEREF, value: (null)
type: T_IDENT, value: byPointer
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: A
type: T_IDENT, value: object
type: T_SEMICOLON, value: (null)
type: T_COMMENT, value: (null)
type: T_IDENT, value: A
type: T_IDENT, value: another
type: T_ASSIGN, value: (null)
type: T_BRACE_L, value: (null)
type: T_INT, value: 1
type: T_COMMA, value: (null)
type: T_COMMENT, value: (null)
type: T_IDENT, value: c
type: T_ASSIGN, value: (null)
type: T_INT, value: 10
type: T_COMMA, value: (null)
type: T_INT, value: 123
type: T_COMMA, value: (null)
type: T_COMMENT, value: (null)
type: T_BRACE_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: return
type: T_CHAR, value: x
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_IDENT, value: void
type: T_PAREN_L, value: (null)
type: T_PAREN_R, value: (null)
type: T_IDENT, value: normalFunction
type: T_ASSIGN, value: (null)
type: T_BRACE_L, value: (null)
type: T_IDENT, value: io
type: T_DOT, value: (null)
type: T_IDENT, value: print
type: T_PAREN_L, value: (null)
type: T_IDENT, value: stuff
type: T_PAREN_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_IDENT, value: i32
type: T_PAREN_L, value: (null)
type: T_PAREN_R, value: (null)
type: T_ASSIGN, value: (null)
type: T_BRACE_L, value: (null)
type: T_IDENT, value: i32
type: T_IDENT, value: a
type: T_ASSIGN, value: (null)
type: T_PAREN_L, value: (null)
type: T_PAREN_L, value: (null)
type: T_INT, value: 10
type: T_PLUS, value: (null)
type: T_INT, value: 5
type: T_PAREN_R, value: (null)
type: T_MULT, value: (null)
type: T_INT, value: 600
type: T_DIV, value: (null)
type: T_INT, value: 4
type: T_PLUS, value: (null)
type: T_INT, value: 8
type: T_PAREN_R, value: (null)
type: T_MULT, value: (null)
type: T_INT, value: 3
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: while
type: T_PAREN_L, value: (null)
type: T_IDENT, value: a
type: T_LT, value: (null)
type: T_INT, value: 0
type: T_PAREN_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_IDENT, value: a
type: T_ASLSHIFT, value: (null)
type: T_INT, value: 1
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_KEYWORD, value: for
type: T_PAREN_L, value: (null)
type: T_IDENT, value: i32
type: T_IDENT, value: i
type: T_ASSIGN, value: (null)
type: T_INT, value: 0
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: i
type: T_LT, value: (null)
type: T_INT, value: 100
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: i
type: T_ASPLUS, value: (null)
type: T_INT, value: 1
type: T_PAREN_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_KEYWORD, value: if
type: T_PAREN_L, value: (null)
type: T_IDENT, value: i
type: T_EQ, value: (null)
type: T_INT, value: 5
type: T_PAREN_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_IDENT, value: continue
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_KEYWORD, value: if
type: T_PAREN_L, value: (null)
type: T_IDENT, value: i
type: T_EQ, value: (null)
type: T_INT, value: 8
type: T_PAREN_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_IDENT, value: break
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_BRACE_R, value: (null)
type: T_IDENT, value: f32
type: T_IDENT, value: f
type: T_ASSIGN, value: (null)
type: T_IDENT, value: sqrt
type: T_PAREN_L, value: (null)
type: T_INT, value: 2
type: T_PAREN_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: f32
type: T_IDENT, value: g
type: T_ASSIGN, value: (null)
type: T_IDENT, value: sqrt
type: T_PAREN_L, value: (null)
type: T_FLOAT, value: 2.5e
type: T_MINUS, value: (null)
type: T_INT, value: 4
type: T_PAREN_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: if
type: T_PAREN_L, value: (null)
type: T_IDENT, value: true
type: T_PAREN_R, value: (null)
type: T_KEYWORD, value: if
type: T_PAREN_L, value: (null)
type: T_IDENT, value: a
type: T_EQ, value: (null)
type: T_INT, value: 10
type: T_PAREN_R, value: (null)
type: T_IDENT, value: a
type: T_ASSIGN, value: (null)
type: T_INT, value: 11
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: else
type: T_IDENT, value: a
type: T_ASSIGN, value: (null)
type: T_INT, value: 10
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: switch
type: T_PAREN_L, value: (null)
type: T_IDENT, value: g
type: T_MULT, value: (null)
type: T_FLOAT, value: 0.5
type: T_PAREN_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_KEYWORD, value: case
type: T_PAREN_L, value: (null)
type: T_INT, value: 0
type: T_PAREN_R, value: (null)
type: T_IDENT, value: print
type: T_PAREN_L, value: (null)
type: T_IDENT, value: g
type: T_PAREN_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_KEYWORD, value: case
type: T_PAREN_L, value: (null)
type: T_INT, value: 1
type: T_COMMA, value: (null)
type: T_INT, value: 2
type: T_COMMA, value: (null)
type: T_INT, value: 3
type: T_PAREN_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_IDENT, value: char
type: T_IDENT, value: x
type: T_ASSIGN, value: (null)
type: T_CHAR, value: 1
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: print
type: T_PAREN_L, value: (null)
type: T_IDENT, value: x
type: T_PAREN_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_KEYWORD, value: case
type: T_PAREN_L, value: (null)
type: T_PAREN_R, value: (null)
type: T_BRACE_L, value: (null)
type: T_IDENT, value: print
type: T_PAREN_L, value: (null)
type: T_STRING, value: default
type: T_PAREN_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_BRACE_R, value: (null)
type: T_BRACE_R, value: (null)
type: T_IDENT, value: char
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_IDENT, value: string
type: T_ASSIGN, value: (null)
type: T_STRING, value: string $#%$^%$^$ \" escape^
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: char
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_IDENT, value: string1
type: T_ASSIGN, value: (null)
type: T_STRING, value: concat stringishere
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: char
type: T_BRACK_L, value: (null)
type: T_BRACK_R, value: (null)
type: T_IDENT, value: longstring
type: T_ASSIGN, value: (null)
type: T_STRING, value: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: A
type: T_IDENT, value: a
type: T_SEMICOLON, value: (null)
type: T_IDENT, value: a
type: T_COLON, value: (null)
type: T_IDENT, value: method
type: T_PAREN_L, value: (null)
type: T_PAREN_R, value: (null)
type: T_SEMICOLON, value: (null)
type: T_COMMENT, value: (null)
type: T_BRACE_R, value: (null)
type: T_COMMENT, value: (null)
type: T_EOT, value: (null)
